### PR TITLE
cpu/native: fix bug in periph_timer

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

While debugging https://github.com/RIOT-OS/RIOT/pull/18977#issuecomment-1764258356 it became obvious that the `periph_timer` in `native` is broken and issues early IRQs. This replaces the use of `setitimer` that cannot use a monotonic clock source with `timer_settime()`.

### Testing procedure

I have some non-publishable code that tests if the time an ISR fires in terms of `timer_read()` is no earlier than the time expected. This occasionally triggered with `master`, but I didn't see any of these issues anymore with this PR. I guess I should revive my PR to spice up the periph timer tests and add a polished version of this and let this run for an hour or two.

The tests ins `tests/periph/timer*` should still succeed on `native`. (They do for me in a container running `riot/riotbuild`).

### Issues/PRs references

Found while debugging https://github.com/RIOT-OS/RIOT/pull/18977#issuecomment-1764258356

### CODE REVIEW

**Comment by Wer-Wolf** [comment]
Could this be because we use ITIMER_REAL as clock source, which is affected by adjtime(2) and settimeofday(2)?
Maybe this problem could be solved by using timer_create() with CLOCK_MONOTONIC.

**Comment by MrKevinWeiss** [comment]
Should this be backported?  Probably not a bad idea.

File: cpu/native/periph/timer.c [hunk 1]
```diff
@@ -26,18 +26,20 @@
...
-#include <sys/time.h>
+#include <err.h>
 #include <signal.h>
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
-#include <err.h>
+#include <sys/time.h>
+#include <time.h>
+#include <time.h>
```
Comment (kaspar030) [inline cpu/native/periph/timer.c:36]:
```suggestion
```

**Comment by maribu** [comment]
> Could this be because we use ITIMER_REAL as clock source, which is affected by adjtime(2) and settimeofday(2)? Maybe this problem could be solved by using timer_create() with CLOCK_MONOTONIC.

Yes, indeed! This was actually also my first intuition. But rather than reading the code I did a quick grep for `CLOCK_[A-Z]*`, which only showed matches for `CLOCK_MONOTONIC` and none `CLOCK_REALTIME`, so I assumed this wasn't the issue. Sometime actually reading the code containing the bug does pay off :sweat_smile:

**Comment by benpicco** [comment]
Hm I still get

```
>>> timeout happened 9739 µs early <<<
>>> timeout happened 9623 µs early <<<
>>> timeout happened 2156 µs early <<<
>>> timeout happened 3265 µs early <<<
>>> timeout happened 9689 µs early <<<
>>> timeout happened 9795 µs early <<<
>>> timeout happened 2262 µs early <<<
>>> timeout happened 9846 µs early <<<
>>> timeout happened 9796 µs early <<<
>>> timeout happened 9647 µs early <<<
>>> timeout happened 4028 µs early <<<
>>> timeout happened 9670 µs early <<<
>>> timeout happened 4650 µs early <<<
>>> timeout happened 9687 µs early <<<
>>> timeout happened 9769 µs early <<<
```

with this (and the patches from #18977) applied

**Comment by Wer-Wolf** [comment]
Maybe the fact that timer_start()/timer_stop() do not stop the value returned by timer_read() from incrementing could also be the reason for this.

**Comment by maribu** [comment]
@benpicco: That's expected. Most of the time that bug triggered it was not due to the `periph_timer` bug. But it should be a tad less frequent now, with one of the causes identified and fixed.

**Review by benpicco — APPROVED** [review]
I'm trusting our timer unittests

**Comment by maribu** [comment]
> Urgh so are you suggesting there is (at least) a _third_ bug

Yes :) But on the other hand 2/3 of the issue is hopefully solved. (Well, the third bug is the one with the most fallout, though).

**Comment by MrKevinWeiss** [comment]
Do we think this can get in the next little while or should we make it a known issue for this release?

**Comment by maribu** [comment]
Murdock found quite the bug. I really should give more priority my rewrite of the `periph_timer` test app higher on my todo, given that the current one was unable to detect the completely broken `timer_stop()` implementation :scream:

**Comment by MrKevinWeiss** [comment]
bors merge

**Comment by chrysn** [comment]
A brief note from a workshop, this broke building on Debian 11.

**Pull request merged**